lib/kernel.js

```diff
@@ -69,62 +270,132 @@ export default class Kernel {
+    const wrappedOnResults = this._wrapExecutionResultsCallback(onResults);
+    this.firstMiddlewareAdapter.execute(
+      code,
+      (message: Message, channel: string) => {
+        wrappedOnResults(message, channel);
+
+        if (
+          channel == "iopub" &&
+          message.header.msg_type === "status" &&
+          message.content.execution_state === "idle"
+        ) {
+          this._callWatchCallbacks();
+        }
+      }
+    );
   }
 
   executeWatch(code: string, onResults: Function) {
-    throw new Error("Kernel: executeWatch method not implemented");
+    this.firstMiddlewareAdapter.execute(
+      code,
+      this._wrapExecutionResultsCallback(onResults)
+    );
   }
 
-  complete(code: string, onResults: Function) {
-    throw new Error("Kernel: complete method not implemented");
+  _callWatchCallbacks() {
+    this.watchCallbacks.forEach(watchCallback => watchCallback());
   }
 
-  inspect(code: string, curorPos: number, onResults: Function) {
-    throw new Error("Kernel: inspect method not implemented");
-  }
+  /*
+   * Takes a callback that accepts execution results in a hydrogen-internal
+   * format and wraps it to accept Jupyter message/channel pairs instead.
+   * Kernels and plugins all operate on types specified by the Jupyter messaging
+   * protocol in order to maximize compatibility, but hydrogen internally uses
+   * its own types.
+   */
+  _wrapExecutionResultsCallback(onResults: Function) {
+    return (message: Message, channel: string) => {
+      if (channel === "shell") {
+        const { status } = message.content;
+        if (status === "error" || status === "ok") {
+          onResults({
+            data: status,
+            stream: "status"
+          });
+        } else {
+          console.warn(
+            "Kernel: ignoring unexpected value for message.content.status"
+          );
+        }
+      } else if (channel === "iopub") {
+        if (message.header.msg_type === "execute_input") {
+          onResults({
+            data: message.content.execution_count,
+            stream: "execution_count"
+          });
+        }
 
-  _parseIOMessage(message: Message) {
-    let result = this._parseExecuteInputIOMessage(message);
+        // TODO(nikita): Consider converting to V5 elsewhere, so that plugins
+        // never have to deal with messages in the V4 format
+        const result = msgSpecToNotebookFormat(msgSpecV4toV5(message));
+        onResults(result);
+      } else if (channel === "stdin") {
+        if (message.header.msg_type !== "input_request") {
+          return;
+        }
 
-    if (!result) {
-      result = msgSpecToNotebookFormat(msgSpecV4toV5(message));
-    }
+        const { prompt } = message.content;
 
-    return result;
+        // TODO(nikita): perhaps it would make sense to install middleware for
+        // sending input replies
+        const inputView = new InputView({ prompt }, (input: string) =>
```

Comment:
Input requests are handled here. (Note that GitHub by default doesn't render the diff for this file because it's too big)